### PR TITLE
Add Gemma4 EAGLE-3 hidden-state taps

### PR DESCRIPTION
**Pull request by digantdesai**
Adds validated tap-layer configuration, a shared decode path that can
return concatenated tap features, and forward_logits_taps for
soft-capped logits with EAGLE-3 features. Includes tests for tap
ordering, validation, logits parity, and last-logits behavior.

Authored with assistance from Claude Code.

### CODE REVIEW

**Comment by digantdesai** [comment]
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* #20349
* #20348
* #20347
* #20346
* #20345
* #20344
* #20157
* #20156
* #20155
* #20154
* #20153
* #20152
* #20151
* __->__ #20150
* #20149

File: examples/models/gemma4_31b/model.py [hunk 1]
```diff
@@ -482,18 +550,41 @@ def forward(
...
+    def set_eagle_tap_layers(self, layers: list) -> None:
+        """Set and validate EAGLE-3 tap layers."""
+        validate_eagle_tap_layers(layers, self.config.num_hidden_layers)
+        self.config.eagle_tap_layers = list(layers)
+
+    def forward_logits_taps(
+        self,
+        tokens: torch.LongTensor,
+        input_pos: torch.LongTensor,
+        last_logits_only: bool = True,
```
Comment (metascroy) [inline examples/models/gemma4_31b/model.py:572]: For tapped model, do we need all logits, not just final so sampling can be done outside model?

If so, why is last_logits_only default to True on forward_logits_taps?

Comment (digantdesai) [reply examples/models/gemma4_31b/model.py:572]: we do need all logits for verification, this is independent of where you do sampling. But not for regular prefill/decode.

File: examples/models/gemma4_31b/model.py [hunk 2]
```diff
@@ -482,18 +550,41 @@ def forward(
...
+        materialize per-position float32 logits over the full vocabulary.
+
+        Returns:
+            logits: (B, 1, vocab_size) soft-capped float32, or (B, T, vocab_size)
+                    when ``last_logits_only=False``.
+            taps:   (B, T, len(eagle_tap_layers) * hidden_size) or None.
+        """
+        x, taps = self._decode(tokens, input_pos, collect_taps=True)
+        if last_logits_only:
+            x = x[:, -1:, :]
```
Comment (mergennachin) [inline examples/models/gemma4_31b/model.py:586]: don't you want to do slicing inside _decode?  i.e., have _decode take last_taps_only boolean parameter

**Review by mergennachin — APPROVED** [review]
